### PR TITLE
#2120 make unit dynamic

### LINKED ISSUE #2120: BOM - Unit Should be More Dynamic

### PR DESCRIPTION
**Pull request by ArmaanMehra**
## Changes

So far what I've done is I've written out the logic for making the units more dynamic. However, I'm working with mock units since there don't seem to be any units when I work with however it tries to access the actual data.
So far:
- Able to parse Units and convert all units with "Pack of..." to a single "Pack" option.
- If Pack option is selected, then a Pack Size option pops up, and all options of pack size from original units datais provided.

## Notes

IMPORTANT NOTE: This issue is not complete yet, I've opened this PR for feedback on what I've done so far and advice on how to move forward.

Rn the issue I'm running into is that I'm not 100% sure if what I'm doing works, since I don't have access to backend Units data when I'm running this on localhost.

## Test Cases

Unsure on how to test this, or if it needs testing since this is all in fronted.

## Screenshots

image
image
image

## To Do

- [ ] Still need to convert the dynamic unit data back to a string to send back to backend.

## Checklist

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #2120

### CODE REVIEW

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/BOM/MaterialForm/MaterialFormView.tsx [hunk 1]
```diff
@@ -55,6 +58,16 @@ const MaterialFormView: React.FC<MaterialFormViewProps> = ({
     setValue(`price`, parseFloat(value.toFixed(2)));
   };
 
+  const [showPackInput, setShowPackInput] = useState(true);
+
+  // Existing logic to watch unit selection
+  const unitName = watch('unitName');
```
Comment (Aaryan1203) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/BOM/MaterialForm/MaterialFormView.tsx:64]: I dont think this is requierd

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/BOM/MaterialForm/MaterialForm.tsx [hunk 2]
```diff
@@ -143,12 +145,34 @@ const MaterialForm: React.FC<MaterialFormProps> = ({ submitText, onSubmit, defau
...
+      if (unit.name.startsWith('Pack of')) {
+        return { ...unit, name: 'Pack', materials: [] };
+      }
+      return unit;
+    })
+    .filter((unit, index, self) => index === self.findIndex((t) => t.name === unit.name));
+
+  const packSizes = mockUnits
+    .filter((unit) => unit.name.startsWith('Pack of'))
+    .map((unit) => unit.name.replace('Pack of ', ''));
```
Comment (Aaryan1203) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/BOM/MaterialForm/MaterialForm.tsx:167]: I think there will be a better way to do this, but i think we can change it once we understand what exactly were doing

**Review by Aaryan1203 — CHANGES_REQUESTED** [review]
I am a little confused with that exactly the ticket is supposed to be doing right now but I did leave a couple comments for now

**Review by RChandler234 — CHANGES_REQUESTED** [review]
code's looking good! Only thing left is to add support for packs in the admin tool for adding units

**Pull request closed without merging**